final_project/snapvisit/models.py

```diff
@@ -1,3 +1,61 @@
 from django.db import models
+from django.contrib.auth.models import AbstractBaseUser, BaseUserManager, PermissionsMixin
+from phonenumber_field.modelfields import PhoneNumberField
+import datetime
 
 # Create your models here.
+
+
+class CustomAccountManager(BaseUserManager):
+
+    def create_superuser(self, email, user_name, password, **other_fields):
+        other_fields.setdefault('is_staff', True)
+        other_fields.setdefault('is_superuser', True)
+        if other_fields.get('is_staff') is not True:
+            raise ValueError(
+                'Superuser must be assigned to is_staff=True.'
+            )
+        if other_fields.get('is_superuser') is not True:
+            raise ValueError(
+                'Superuser must be assigned to is_superuser=True.'
+            )
+        return self.create_user(email, user_name, password, **other_fields)
+
+    def create_user(self, email, user_name, password, **other_fields):
+        if not email:
+            raise ValueError("Users must have an email address.")
+        if not user_name:
+            raise ValueError("Users must have an username.")
+        email = self.normalize_email(email)
+        user = self.model(
+            email=email,
+            user_name=user_name,
+            **other_fields)
+        user.set_password(password)
+        user.save()
+        return user
+
+
+class Profile(AbstractBaseUser, PermissionsMixin):
```

Comment:
I think the AbstractUser model should be used instead of AbstractBaseUser. AbstractUser has most of the fields already declared, what's more, it will improve our work in the future when we want to extend this application. What do you think?

Reply:
AbstractUser dont give a chance to change authentication system. I choice AbstractBaseUser to have full control about User model and what most important - change method of login from "username and password" to "email and password". Most of sites now used to email and everybody have easier to remind email instead of username. Should I merge?

Reply:
Ok i just merge it to push work ahead!